### ISSUE TITLE
Diff edits seem to not work at all

### ISSUE THREAD

**Issue by Green0-0** [cline/cline#1010]
### What happened?

I have attached an image showing my problem. 
I have created a simple python file that needs to be diff edited:
```python
# This is a sentence
# This is a sentence
# This is a sentence
# This is a sentence
# This is a sentence
# This is a sentence
# This is a sentence
# This is a sentence
# Diff edit this line into a print statement 
# This is a sentence
# This is a sentence
# This is a sentence
# This is a sentence
# This is a sentence
# This is a sentence
# This is a sentence
# This is a sentence
```

And I have asked the model (gemini-2.0-flash-exp) to diff edit it. It repeatedly fails. I understand that I am not using Claude 3.5 Sonnet, but given how simple the task is, it seems rather strange that the model cannot successfully do it. Moreover, the search block it has provided (attached below) seems like it should match something:

```
The SEARCH block:
# Diff edit this line into a print statement
...does not match anything in the file.
```

I have not once gotten the diff edit to work. It is possible that somehow, the tool call is simply misleading gemini-2.0-flash-exp into adding an extra whitespace or newline at the end. But this doesn't really make sense, and it isn't that gemini-2.0-flash is stupid either. If that is however the problem, it should be quite easy to fix with custom instructions.

### Steps to reproduce

1. Create a python file called diff_edit_test.py. Add the content:
```python
# This is a sentence
# This is a sentence
# This is a sentence
# This is a sentence
# This is a sentence
# This is a sentence
# This is a sentence
# This is a sentence
# Diff edit this line into a print statement 
# This is a sentence
# This is a sentence
# This is a sentence
# This is a sentence
# This is a sentence
# This is a sentence
# This is a sentence
# This is a sentence
```
2. Send cline/gemini-2.0-flash-exp the prompt "Diff edits are currently broken, and I need to find out why. Read the diff_edit_test.py file and try to diff edit line 9 (# Diff edit this line into a print statement ) into print("diff edits work!")"

### Relevant API REQUEST output

```shell
replace_in_file for 'diff_edit_test.py'] Result:

The tool execution failed with the following error:
<error>
Error writing file: The SEARCH block:
# Diff edit this line into a print statement
...does not match anything in the file.
This is likely due to differences in whitespace or line endings between the SEARCH block and the actual file. Try again with a more precise SEARCH block.
(If you keep running into this error, you may use the write_to_file tool as a workaround.)
</error>
```

### Additional context

![image](https://github.com/user-attachments/assets/cbb449e4-eabd-4de6-9cfa-2084a423b52f)

**Comment by spyhunk** [comment]
Yup! just watching this same bug occurance on loop. I also provided Custom Instruction: "Don't use replace_in_file to edit files, you are only allowed to use write_to_file"

**Comment by MBrekhof** [comment]
I have the same problem on .razor files (Blazor), I just disabled the diff editing stuff.

**Comment by GLStephen** [comment]
I see similar behavior and I think it's due to Prettier running after save and changing whitespace. I --think--.

**Comment by pablotoledo** [comment]
I had to downgrade to 2.2.0 due to similar issues with Ollama Models like qwen

**Comment by saoudrizwan** [comment]
Can I please see the actual search/replace provided by the model? You can find this by expanding the file:
<img width="340" alt="image" src="https://github.com/user-attachments/assets/3cd060d7-1050-4468-9d87-e472a5ee6d2c" />

This doesn't seem to be a model problem based on the information you provided. If it's providing a valid search, and the extension is failing to apply the diff then it is probably an issue with parsing the model's diff properly.

**Comment by Green0-0** [comment]
I cannot open the ``diff edit failed`` block to see the search/replace, and there is no ``cline wants to edit this file`` where it is supposed to be. Instead, I have provided this, which should be taken with a grain of salt as it is AI generated, but is likely correct:
![image](https://github.com/user-attachments/assets/fc3f5f41-d2e3-4b8f-b9a4-514e21770d81)

**Comment by saoudrizwan** [comment]
Oh wow it seems it's unable to even call the tool correctly in the first place. Note to contributors: try to reproduce this with gemini and check global storage (lib/app support/code/saoudrizwan.claude-dev/user/global/tasks/conversation_history.json) and see what the model is outputting that the extension is failing to parse. It could be introducing some artifact we need to filter.

**Comment by xueshuai0922** [comment]
I had the same problem via deepseek v3

![image](https://github.com/user-attachments/assets/b99c479f-e411-41bb-be98-72cae3d23796)

**Comment by xueshuai0922** [comment]
> I have the same problem on .razor files (Blazor), I just disabled the diff editing stuff.我在 .razor 文件 （Blazor） 上遇到了同样的问题，我刚刚禁用了差异编辑内容。

how ? in cline settings ?

**Comment by Abelhubprog** [comment]
Diff edit still a problem, is downgrading resolves this?

**Comment by yo-chanan** [comment]
This is still a problem and makes the extension totally unusable. It's happening 90% of the time.

**Comment by celestial-vault** [comment]
Thanks for reporting! This diff editing/tool calling failure with local models including Ollama has been consolidated into #4362 - Improve Ollama Local Model Compatibility.

**Workarounds:**
- Use tool-calling optimized models: `hhao/qwen2.5-coder-tools` or `acidtib/qwen2.5-coder-cline:7b`
- Add custom instruction: "Don't use replace_in_file to edit files, you are only allowed to use write_to_file"
- Consider LM Studio as alternative for better tool calling support

The master issue addresses tool calling training deficiencies affecting local models including Ollama, Qwen, and DeepSeek.

**Comment by yo-chanan** [comment]
This issue has nothing to do with local models, and the models you list are not even the one he mentions (Gemini). Moreover, we already have a checkbox to disable replace_in_file, but write_to_file can't handle if the file is even a little large.
